lib/puppet/provider/foreman_hostgroup/rest_v3.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+Puppet::Type.type(:foreman_hostgroup).provide(
+  :rest_v3,
+  parent: Puppet::Type.type(:foreman_resource).provider(:rest_v3)
+) do
+  confine feature: %i[json oauth]
+
+  def initialize(value = {})
+    super(value)
+    @property_flush = {}
+  end
+
+  def hostgroup
+    @hostgroup ||= begin
+      path = 'api/v2/hostgroups'
+      search_name = resource[:name]
+      search_title = if resource[:parent_hostgroup]
+                       "#{resource[:parent_hostgroup]}/#{resource[:name]}"
+                     else
+                       search_name
+                     end
+      Puppet.debug("Searching for hostgroup with name #{search_name} and title #{search_title}")
+      r = request(:get, path, search: %(title="#{search_title}" and name="#{search_name}"))
+
+      raise Puppet::Error, "Error making GET request to Foreman at #{request_uri(path)}: #{error_message(r)}" unless success?(r)
+
+      results = JSON.parse(r.body)['results']
+      unless results.empty?
+        raise Puppet::Error, "Too many hostgroups found when looking for hostgroup with name #{search_name} and title #{search_title}" if results.size > 1
+
+        get_hostgroup_by_id(results[0]['id'])
+      end
+    end
+  end
+
+  def get_hostgroup_by_id(id)
+    path = "api/v2/hostgroups/#{id}"
+    r = request(:get, path)
+
+    return JSON.parse(r.body) if success?(r)
+
+    raise Puppet::Error, "Error making GET request to Foreman at #{request_uri(path)}: #{error_message(r)}"
+  end
+
+  def id
+    hostgroup ? hostgroup['id'] : nil
+  end
+
+  def exists?
+    !id.nil?
+  end
+
+  def parent_hostgroup
+    return nil unless resource[:parent_hostgroup]
+
+    @parent_hostgroup ||= begin
+      path = 'api/v2/hostgroups'
+      search_title = resource[:parent_hostgroup]
+      search_name = if resource[:parent_hostgroup_name]
+                      resource[:parent_hostgroup_name]
+                    elsif %r{^(?<_parent>.*)/(?<name>.*)$} =~ resource[:parent_hostgroup]
+                      name
```

Comment:
From https://medium.com/rubycademy/3-ways-to-use-regexp-capture-groups-with-back-references-in-ruby-b4969cc9b3ec
> The named capture feature, when it’s used with the =~ operator, has a powerful mechanism that automatically assigns the result of a matching group into a local variable using the name assigned to the group into the regular expression

Reply:
TIL. Perhaps a comment can help there. It may not be obvious to everyone. Another less magical way may be:
```ruby
elsif resource[:parent_hostgroup].include?('/')
  resource[:parent_hostgroup].split('/').last
```

Actually, that would also work if there is no `/` and since `search_title` is `resource[:parent_hostgroup]` it could be simplified to just
```ruby
else
  search_title.split('/').last
end
```

Which can then mean you can reduce it to:
```ruby
search_name = resource[:parent_hostgroup_name] || search_title.split('/').last
```